### ISSUE TITLE
Custom Widget type not assignable to CmsFieldBase - Typescript

### ISSUE THREAD

**Issue by chrisleyva** [netlify/netlify-cms#6527]
Hello, team. Thank you for your hard work on this package.

Following the docs, I created a custom widget: `videoEmbed`.
But when I try to add it as a field in one of my collections of files, of course it's not an acceptable widget type according to `netlify-cms-core`'s shipped typings.

**To Reproduce**
1. Create custom widget & register with the CMS
2. Create a `config.ts` file which specifies the `CmsConfig` object which is passed into `CMS.init(config)`
3. In the CmsConfig object, add a File Collection with a single file with a single file/page.
4. In that file's `fields` , add the custom widget (in my case `videoEmbed`)
5. Notice the Typescript error (screenshots)

![image](https://user-images.githubusercontent.com/1437143/182423276-9976abff-6303-41bf-ac2b-e8754c791edd.png)

## TLDR: Expected Behavior
I would expect a typescript-safe way to specify custom widgets in a typescript config file. Maybe there is?
Is there a "nice" and sanctioned way to add type support for custom widgets in `cmsConfig.ts`?

## Workaround 1
Add `as unknown as CmsField` after each custom widget field like so:
```typescript
{
  label: 'Video',
  name: 'videoEmbed',
  widget: 'videoEmbed',
} as unknown as CmsField,
```
This "works" as you can see in the screenshot: 
![image](https://user-images.githubusercontent.com/1437143/182423534-8d84ce2f-609e-47de-9e7a-5664ef41e18e.png)

### Downside
The more I use custom widgets on more and more pages, the more I have to repeat the use of `as unknown as CmsField`. This doesn't feel right.

Of course, I could just create a single const and use that instead but I _still_ don't think this is the right way to go:
```typescript
const videoEmbedCmsField = {
  label: 'Video',
  name: 'videoEmbed',
  widget: 'videoEmbed',
} as unknown as CmsField;

// ...
{
  label: 'Intro',
  name: 'intro',
  widget: 'markdown',
},
videoEmbedCmsField,
{
  label: 'Show Gallery?',
  name: 'showGallery',
  widget: 'boolean',
},
// ...
```

## Workaround 2 (need help here)
The other approach is to modify the `CmsConfig` type _aaaaaaall_ the way until I reach the `field: CmsField` type so that I can add `{ widget: 'videoEmbed' }` as an acceptable widget type.

This **feels** like the right way to do it **but** it's very cumbersome AND, more importantly, I haven't been able to figure out the right combination of type unions & intersections. I could use some help here.

This screenshot shows my attempt but it doesn't work; I still can't use my custom widget (the error is in the next screenshot)
![image](https://user-images.githubusercontent.com/1437143/182428062-ef514241-8034-42dd-85d3-7dcaf18219d7.png)

The error `Type '{ label: string; name: string; widget: "videoEmbed"; }' is not assignable to type '(CmsFieldBase & CmsFieldBoolean) | ...`:
![image](https://user-images.githubusercontent.com/1437143/182426008-ee9ab9f1-05b5-4b1b-86b7-68f694a809ef.png)

## Workaround 3
Modify `type CmsField` in `netlify-cms-core/index.d.ts` to include the new widget type:
```typescript index.d.ts
  export interface CmsFieldVideoEmbed {
    widget: 'videoEmbed';
    default?: string;
  }
  
  export type CmsField = CmsFieldBase &
    (
      | CmsFieldBoolean
      | CmsFieldCode
      | CmsFieldColor
      | CmsFieldDateTime
      | CmsFieldFileOrImage
      | CmsFieldList
      | CmsFieldMap
      | CmsFieldMarkdown
      | CmsFieldNumber
      | CmsFieldObject
      | CmsFieldRelation
      | CmsFieldSelect
      | CmsFieldHidden
      | CmsFieldStringOrText
      | CmsFieldMeta
      | CmsFieldVideoEmbed // <-- my custom widget type
    );
```

Obviously, this is not the right way to do it **but** it allows my project to run locally. 🎉 

# Conclusion

Is there a proper way to do this? Thank you for your time.

### Applicable Versions
- netlify-cms-app 2.15.72
- netlify-cms-core 2.55.2
 - Git provider: GitHub, BitBucket
 - OS: macOS Monterey 12.5
 - Browser version Chrome 103
 - Node.JS version: v16.10.0

## CMS configuration
```typescript
{
  backend: {
    name: 'git-gateway',
    branch: 'main',
    commit_messages: {
      create: 'docs: create {{collection}} “{{slug}}”',
      update: 'docs: update {{collection}} “{{slug}}”',
      delete: 'docs: delete {{collection}} “{{slug}}”',
      uploadMedia: 'docs: upload “{{path}}”',
      deleteMedia: 'docs: delete “{{path}}”',
      openAuthoring: 'docs: {{message}}',
    },
  },
  local_backend: true,
  media_folder: '/public/uploads',
  public_folder: '/uploads',
  collections: [
    {
      label: 'Pages',
      name: 'pages',
      files: [
        {
          label: 'Home',
          name: 'home',
          file: 'content/home.md',
          fields: [
            {
              label: 'Title',
              name: 'title',
              widget: 'string',
            },
            {
              label: 'Intro',
              name: 'intro',
              widget: 'markdown',
            },
            {
              label: 'Show Gallery?',
              name: 'showGallery',
              widget: 'boolean',
            },
            {
              label: 'Video',
              name: 'videoEmbed',
              widget: 'videoEmbed', // <-- This is the issue here
            },
            {
              label: 'Features',
              name: 'features',
              widget: 'list',
              fields: [
                { label: 'Title', name: 'title', widget: 'string' },
                {
                  label: 'Content',
                  name: 'content',
                  widget: 'markdown',
                  modes: ['rich_text', 'raw'],
                },
              ],
            },
            {
              label: 'Gallery Images',
              name: 'gallery',
              widget: 'list',
              label_singular: 'Image',
              add_to_top: true,
              fields: [{ label: 'Image', name: 'image', widget: 'image' }],
            },
          ],
        },
      ],
    },
  ],
}
```
